### PR TITLE
no-dupe-else-if

### PR DESCRIPTION
**Pull request by magurotuna**
~This is WORK IN PROGRESS~

This PR adds [no-dupe-else-if - Rules - ESLint - Pluggable JavaScript linter](https://eslint.org/docs/rules/no-dupe-else-if) rule.

The original implementation is [here](https://github.com/eslint/eslint/blob/master/lib/rules/no-dupe-else-if.js), and it's so complicated.
I did my best to port the original implementation, but the code got dirty, although all the tests are passed.

Also I'm concerned about bad time complexity and cloning `Vec` many times, which may cause bad performance.

### CODE REVIEW

**Comment by bartlomieju** [comment]
@disizali please review when you get the time

File: src/rules/no_dupe_else_if.rs [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+      "if (a) {} else if (b) {} else if (c && (a || d && b)) {}",
+      34,
+    );
+    assert_lint_err::<NoDupeElseIf>("if (a) {} else if (a || a) {}", 19);
+    assert_lint_err::<NoDupeElseIf>("if (a || a) {} else if (a || a) {}", 24);
+    assert_lint_err::<NoDupeElseIf>("if (a || a) {} else if (a) {}", 24);
+    assert_lint_err::<NoDupeElseIf>("if (a) {} else if (a && a) {}", 19);
+    assert_lint_err::<NoDupeElseIf>("if (a && a) {} else if (a && a) {}", 24);
+    assert_lint_err::<NoDupeElseIf>("if (a && a) {} else if (a) {}", 24);
+  }
```
Comment (bartlomieju) [inline src/rules/no_dupe_else_if.rs:421]: One more request from me: please add a test case with nested if:
```
if (something) {
  if (a == 1) {
  } else if (a == 2) {}
}
```

File: src/rules/no_dupe_else_if.rs [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+    (Paren(ParenExpr { ref expr, .. }), _) => equal_in_if_else(&**expr, expr2),
+    (_, Paren(ParenExpr { ref expr, .. })) => equal_in_if_else(expr1, &**expr),
+    (This(_), This(_))
+    | (Array(_), Array(_))
+    | (Object(_), Object(_))
+    | (Fn(_), Fn(_))
+    | (Unary(_), Unary(_))
+    | (Update(_), Update(_))
+    | (Bin(_), Bin(_))
+    | (Assign(_), Member(_))
```
[the remark this answers is not in the document]
Comment (magurotuna) [reply src/rules/no_dupe_else_if.rs:174]: Oops, I made a mistake! I'll fix it soon.

**Comment by magurotuna** [comment]
@disizali @bartlomieju 
Thanks for the review. I've added more tests and fixed the implementation in `equal_in_if_else`.
Could you review again please.

**Pull request merged**